### ISSUE TITLE
How to use serve-index behind a reverse proxy ?

### ISSUE THREAD

**Issue by alemarre** [expressjs/serve-index#53]
Hi,

I'd like to use `serve-index` behind a reverse proxy that adds a root context, i. e : 
`mydomain.com/root-context/my-folder` --> `localhost:3000/my-folder`

The `serve-static` module works fine (if I enter the full adresse by hand, it is correctly resolved), but URLs generated by `serve-index` are absolute URLs, and are not taking in account the root context.

Example of url generated by `serve-index`: 
`mydomain.com/my-folder/other` (should be `mydomain.com/root-context/my-folder/other`)

To summarize : I can access my content with direct URLs, but I can't navigate through `serve-index`'s interface.

- Why using absolute URLS and not relative ones ? For instance, the Apache module uses relative URLs when serving directory files list.
- If absolutes URLs are absolutely needed, how can I configure a root context ?

Thanks for your help

**Comment by dougwilson** [comment]
Hi @alemarre, the reason this module has historically used absolute URIs in the links is because this module will respond with or without a trailing slash. For example, if you loaded `mydomain.com/root-context/my-folder` in your web browser, then a link to `image.png` would end up going to `mydomain.com/root-context/image.png`, not `mydomain.com/root-context/my-folder/image.png`, since the relative link it relative within the `root-context` folder since there is no trailing slash in the URL.

The way to resolve this is that you need to get your `req.originalUrl` set correctly before entering this module. The best way is to get it setup in Express at the top, so anything in your application trying to generate URLs or even HTTP redirects can do so using the full URL when necessary.

Depending on how your proxy works, you could possibly have the proxy add a request header `X-Original-Url` to the backend request and then add the following to the top of your Express application:
```js
app.use(function (req, res, next) {
  req.originalUrl = req.headers['x-original-url'] || undefined
  next()
})
```

**Comment by alemarre** [comment]
Thanks a lot for your great explanation

**Comment by dougwilson** [comment]
No problem, @alemarre!
